Improve error handling for UI actions using `NotebookOperation`

oturpe opened the pull request:
Fixes issue #2382.

## Fix PageExistsError and PageReadOnlyError inheritance

Classes `PageExistsError` and `PageReadOnlyError` were inherited directly from `zim.Error`,
yet their implementations only make sense as subclasses of `PageError`. Make it so.

Previously, the error message for these two errors was just the page name.
After this fix, the page name is prefixed with _Page already exists_ and Can not modify page respectively.
This makes the error messages much more understandable.

## Propagate exceptions from progress dialogs

`NotebookOperation` performs its work inside the GTK main loop.
Any unhandled exceptions occuring within the main loop are processed within the main loop
and do not propagate to code that started operation.
To allow informing the calling side about exceptions, `NotebookOperation` has attribute exception,
which together with the cancelled is filled by `NotebookOperation`'s top level exception handler.

However, for many UI actions that use NotebookOperation wrapped in a `ProgressDialog`,
the exception attribute is not considered at all.
For these dialogs, the user does not see the expected error dialog should an exception occur.

Fix by re-raising any exception reported by NotebookOperation in the following actions:

* MovePageDialog
* ExportDialog
* check_and_update_index, which has its own menu item, and is also  started in multiple other UI actions

## Avoid unhandled exceptions in `NotebookOperation`

`NotebookOperation` is intended to be executed in the GTK main loop via `run_on_idle()`.
Even if such execution terminates in an unhandled exception, execution of Zim still continues.
Thus, normally the effect of an unhandled exception is to print an error message and a traceback in Zim's debug log.

However, if `sys.excepthook` is overridden, different action may result.
In particular, in Fedora Linux, Automatic Bug Reporting Tool (ABRT)'s Python plugin overrides `sys.excepthook`
to notify the user about unhandled exception in Zim,
and also to upload a crash report to Fedora Analytics Framework.
Since multiple `NotebookOperation` subclasses use exceptions to communicate failures that are preventable by the user,
such as trying to create a page that already exists,
this is not desirable.

To avoid triggering `sys.excepthook`,
stop re-raising exceptions in the already existing `NotebookOperation` catch-all exception handler.
Just print them to the debug log instead.

Note that since the catch-all handler already communicates the exceptions to `NotebookOperation` caller via the `cancelled` and `exception` properties,
it could be argued that all handling of exceptions, even logging them, should be left to the caller.
However, there are places where `NotebookOperation` subclasses are executed,
but those properties are not checked afterwards at all,
and thus the *only* error handling they already have is the debug log.
Since the author of this change does not feel comfortable adding exception handling to all those places,
the solution is to keep logging the exception inside `NotebookOperation`.

Since the error logging is now done explicitly via Python's logging
mechanism, tests also need to be modified to expect the new log rows.

## Also some typo fixes in comments

oturpe committed:
Propagate exceptions from progress dialogs

`NotebookOperation` performs its work inside the GTK main loop. Any
unhandled exceptions occuring within the main loop are processed within
the main loop and do not propagate to code that started operation. To
allow informing the calling side about exceptions, `NotebookOperation`
has attribute `exception`, which together with the `cancelled`
is filled by `NotebookOperation`'s top level exception handler.

However, for many UI actions that use `NotebookOperation` wrapped in
a `ProgressDialog`, the `exception` attribute is not considered at all.
For these dialogs, the user does not see the expected error dialog
should an exception occur.

Fix by re-raising any exception reported by `NotebookOperation` in the
following actions:

* MovePageDialog
* ExportDialog
* check_and_update_index, which has its own menu item, and is also
  started in multiple other UI actions

oturpe committed:
Avoid unhandled exceptions in `NotebookOperation`

`NotebookOperation` is intended to be executed in the GTK main loop via
`run_on_idle()`. Even if such execution terminates in an unhandled
exception, execution of Zim still continues. Thus, normally the effect
of an unhandled exception is to print an error message and a traceback
in Zim's debug log.

However, if `sys.excepthook` is overridden, different action may
result. In particular, in Fedora Linux, Automatic Bug Reporting Tool
(ABRT)'s Python plugin overrides `sys.excepthook` to notify the user
about unhandled exception in Zim, and also to upload a crash report to
Fedora Analytics Framework. Since multiple `NotebookOperation`
subclasses use exceptions to communicate failures that are preventable
by the user, such as trying to create a page that already exists, this
is not desirable.

To avoid triggering `sys.excepthook`, stop re-raising exceptions in
the already existing `NotebookOperation` catch-all exception handler.
Just print them to the debug log instead.

Note that since the catch-all handler already communicates
the exceptions to `NotebookOperation` caller via the `cancelled` and
`exception` properties, it could be argued that all handling of
exceptions, even logging them, should be left to the caller. However,
there are places where `NotebookOperation` subclasses are executed, but
those properties are not checked afterwards at all, and thus the *only*
error handling they already have is the debug log. Since the author of
this change does not feel comfortable adding exception handling to all
those places, the solution is to keep logging the exception inside
`NotebookOperation`.

Since the error logging is now done explicitly via Python's logging
mechanism, tests also need to be modified to expect the new log rows.

oturpe commented:
I have updated content and description of this pull request.

introt reviewed (approved):
Looks good to me, thanks!

oturpe committed:
Fix some typos in comments

oturpe commented:
Since I realized that for tests, there is a serious conflict with this merge request and #2490, I decided to close the other one and merge its content here. The merged content is in a separate commit. Sorry for the noise.

jaap-karssenberg commented:
I agree that the error handling in the operations makes sense this way.

Want to merge this path, but feel the refactoring of the location of the PageNotFoundError(s) is a bit of noise - refactor of PageError itself makes sense, but the NotFound errors only make sense in the notebook context and therefore should belong in the notebook module.

oturpe commented:
Thank you for the review. I'll revert the move of  `PageError` subclasses from `notebook.py` to `page.py`.

oturpe committed:
Fix PageExistsError and PageReadOnlyError inheritance

Classes PageExistsError and PageReadOnlyError were inherited
directly from zim.Error, yet their implementations only make sense as
subclasses of PageError. Make it so. BEcause PageReadOnlyError is
defined in `page.py`, PageError is also moved there and imported from
higher level `notebook.py` that defines the other page errors.

Previously, the error message for these two errors was just the page
name. After this fix, the page name is prefixed with _Page already
exists_ and Can not modify page respectively. This makes the error
messages much more understandable.

oturpe commented:
@jaap-karssenberg I have applied the simplification you proposed. Please review again. CI error messages are about GSpell, so I do not think they are about this pull request. When I ran the test suite locally, it passed (with 21 skipped tests).

jaap-karssenberg commented:
Hi @oturpe ,

Thanks for the update!  I've merged your commits in the develop branch.

Regards,

Jaap
